### PR TITLE
Keypoints (Human Pose Estimate)

### PR DESCRIPTION
**Pull request by StevenBorkman**
# Peer Review Information:
Added support for keypoint labeling and keypoint randomization.

## Editor / Package versioning:
**Editor Version Target (i.e. 19.3, 20.1)**: 2019.3

## Dev Testing:
**Tests Added**: 

**Core Scenario Tested**: 

**At Risk Areas**: 

**Notes + Expectations**: 

## Checklist
- [x] - Updated docs
- [x] - Updated changelog

### CODE REVIEW

File: com.unity.perception/Documentation~/Schema/Synthetic_Dataset_Schema.md [hunk 1]
```diff
@@ -196,36 +196,77 @@ We follow the OpenCV 2D coordinate [system](https://github.com/vvvv/VL.OpenCV/wi
...
+    y:           <float> -- The y coordinate
+    z:           <float> -- The z coordinate
+    w:           <float> -- The w coordinate
+  }     
+  velocity {             -- [Optional] 3d bounding box velocity in meters per second.
+    x:           <float> -- The x coordinate
+    y:           <float> -- The y coordinate
+    z:           <float> -- The z coordinate
+  } 
+  acceleration {         -- 3d bounding box acceleration in meters per second^2.
```
Comment (JonathanHUnity) [inline com.unity.perception/Documentation~/Schema/Synthetic_Dataset_Schema.md:240]: also optional?

File: com.unity.perception/Documentation~/Schema/Synthetic_Dataset_Schema.md [hunk 2]
```diff
@@ -196,36 +196,77 @@ We follow the OpenCV 2D coordinate [system](https://github.com/vvvv/VL.OpenCV/wi
...
+    x:           <float> -- The x coordinate
+    y:           <float> -- The y coordinate
+    z:           <float> -- The z coordinate
+  } 
 }
 ```
+##### Keypoints
 
+Keypoints data (commonly used for human pose estimation). A keypoint capture is associated to a template that defines the keypoints (see annotation.definition file).
```
Comment (JonathanHUnity) [inline com.unity.perception/Documentation~/Schema/Synthetic_Dataset_Schema.md:249]:
```suggestion
Keypoint data, commonly used for human pose estimation. A keypoint capture is associated to a template that defines the keypoints (see annotation.definition file).
```

File: com.unity.perception/Documentation~/Schema/Synthetic_Dataset_Schema.md [hunk 3]
```diff
@@ -305,25 +346,43 @@ Some special cases like semantic segmentation might assign additional values (e.
...
 
 # label enumeration spec, used for annotations like bounding box 2d. This might be a subset of all labels used in simulation.
 annotation_definition.spec {
-  label_id:    <int> -- Integer identifier of the label
-  label_name:  <str> -- String identifier of the label
+  label_id:          <int>           -- Integer identifier of the label
+  label_name:        <str>           -- String identifier of the label
+}
+
+# keypoint template used to define the keypoints and skeletal connections captured by the keypoint labeler.
```
Comment (JonathanHUnity) [inline com.unity.perception/Documentation~/Schema/Synthetic_Dataset_Schema.md:370]: This single big code quote is a bit hard to follow. Maybe break it up into different sections? Then you can also link to the annotation definition section from above.

File: com.unity.perception/Runtime/Randomization/Randomizers/RandomizerExamples/Randomizers/AnimationRandomizer.cs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+                var playable = cached.playables[clip.name];
+                cached.output.SetSourcePlayable(cached.playables[clip.name]);
+                var t = m_FloatParameter.Sample() * clip.length;
+
+                playable.SetTime(t);
+                playable.SetSpeed(0);
+
+                cached.graph.Play();
+
+                cached.poseState.poseState = animationPoseLabel.GetPoseAtTime(t);
```
Comment (JonathanHUnity) [inline com.unity.perception/Runtime/Randomization/Randomizers/RandomizerExamples/Randomizers/AnimationRandomizer.cs:94]: Didn't we decide to separate the randomization from the labeling?

File: com.unity.perception/Runtime/Randomization/Randomizers/RandomizerExamples/Randomizers/ConstantTransformRandomizer.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+namespace UnityEngine.Experimental.Perception.Randomization.Randomizers.SampleRandomizers
+{
+    /// <summary>
+    /// Sets a gameobject tagged with <see cref="ConstantTransformRandomizerTag"/> to the same transform(location, rotation, scale)
+    /// each frame. This randomizer is useful to reset a scene to a start state, especially when used in conjunction
+    /// with other randomizers that may move a gameobject.
+    /// </summary>
+    [Serializable]
+    [AddRandomizerMenu("Perception/Constant Transform Randomizer")]
+    public class ConstantTransformRandomizer : Randomizer
```
Comment (JonathanHUnity) [inline com.unity.perception/Runtime/Randomization/Randomizers/RandomizerExamples/Randomizers/ConstantTransformRandomizer.cs:13]: I assume this is for the animation randomizer? On Nest we turned off root motion on the animations to address this (at least partially) https://docs.unity3d.com/Manual/RootMotion.html

Maybe we want to just add this feature of saving off the transform and resetting at next randomization to the animation randomizer? At the moment this randomizer feels like it will be a bit of a pain to use.

Comment (JonathanHUnity) [reply com.unity.perception/Runtime/Randomization/Randomizers/RandomizerExamples/Randomizers/ConstantTransformRandomizer.cs:13]: I see you already have the root motion option! Makes sense

**Review by JonathanHUnity — CHANGES_REQUESTED** [review]
I still need to play with it, but left a few comments here for now. I think we should also have some specific documentation on the labeler and how to use it.

File: com.unity.perception/Runtime/Randomization/Randomizers/RandomizerExamples/Randomizers/AnimationRandomizer.cs (hunk 4 above)
Comment (StevenBorkman) [reply com.unity.perception/Runtime/Randomization/Randomizers/RandomizerExamples/Randomizers/AnimationRandomizer.cs:94]: Well, I wanted to go down that road but I talked to some folks on the animation team about trying to figure out the current clip and frame from the executing animation, and they really thought that was a bad idea. So I did go away from using the tag to store the info and just added a "cache" data item to the game object that had the labeling. But I definitely could change that so that we are just caching the current clip and time in the cache data, and put the labeling info back on to the labeler. That's probably the best idea.

File: com.unity.perception/Runtime/Randomization/Randomizers/RandomizerExamples/Randomizers/ConstantTransformRandomizer.cs (hunk 5 above)
Comment (StevenBorkman) [reply com.unity.perception/Runtime/Randomization/Randomizers/RandomizerExamples/Randomizers/ConstantTransformRandomizer.cs:13]: Yeah, figured that out later. I think I'm going to remove this randomizer.

File: com.unity.perception/Runtime/GroundTruth/Labelers/AnimationPoseLabel.cs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    /// <summary>
+    /// Record that maps a pose to a timestamp
+    /// </summary>
+    [Serializable]
+    public class PoseTimestampRecord
+    {
+        /// <summary>
+        /// The duration to use for this timestamp in seconds
+        /// </summary>
+        public float duration;
```
Comment (sleal-unity) [inline com.unity.perception/Runtime/GroundTruth/Labelers/AnimationPoseLabel.cs:16]: My suggestion would be to change this duration value to something like startTime or startingKeyFrame. Users would first author a Pose timestamp that starts at 0 then add timestamps with increasing time values at each animation transition point.

**Review by sleal-unity — CHANGES_REQUESTED** [review]

File: com.unity.perception/Runtime/Randomization/Randomizers/RandomizerExamples/Randomizers/AnimationRandomizer.cs (hunk 4 above)
Comment (StevenBorkman) [reply com.unity.perception/Runtime/Randomization/Randomizers/RandomizerExamples/Randomizers/AnimationRandomizer.cs]: Ok, this has been reworked and I think it is more aligned to what you are looking for

File: com.unity.perception/Runtime/Randomization/Randomizers/RandomizerExamples/Randomizers/ConstantTransformRandomizer.cs (hunk 5 above)
Comment (StevenBorkman) [reply com.unity.perception/Runtime/Randomization/Randomizers/RandomizerExamples/Randomizers/ConstantTransformRandomizer.cs]: I have removed this class

**Comment by JonathanHUnity** [comment]
Great! Two more things I think we should add:

* A manual page (or two?) to help users know how to use both the labeling and the randomization. We could really use a sample as well, but that can come later
* At least one new entry on the test rail describing how to test and what asset to use

**Review by sleal-unity — APPROVED** [review]

**Review by JonathanHUnity — APPROVED** [review]
